lib/Conversion/TorchOnnxToTorch/OnnxLstmExpander.cpp

```diff
@@ -0,0 +1,524 @@
+ *         super().__init__()
+ *         self.lstm = torch.nn.LSTM(10, 20, 1)
+ *     @export
+ *     @annotate_args([
+ *         None,
+ *         ([5, 1, 10], torch.float32, True),
+ *         ([1, 1, 20], torch.float32, True),
+ *         ([1, 1, 20], torch.float32, True),
+ *     ])
+ *     def forward(self, input, h0, c0):
+ *         return self.lstm(input, (h0, c0))
+ *
+ * @register_test_case(module_factory=LSTMModule)
+ * def LSTMModule_basic(module, tu: TestUtils):
+ *     inputs = torch.zeros(5,1,10)
+ *     h0 = torch.zeros(1,1,20)
+ *     c0 = torch.zeros(1,1,20)
+ *
+ *     output, (hn, cn) = module.forward(inputs, h0, c0)
+ * @endcode
+ *
+ * @param binder The OpBinder object used for binding operands.
+ */
+LogicalResult OnnxLstmExpander(OpBinder binder,
+                               ConversionPatternRewriter &rewriter) {
+  Location loc = binder.getLoc();
+  mlir::ImplicitLocOpBuilder b(loc, rewriter);
+
+  std::string direction;
+
+  ValueTensorType yTy, Y_hType, Y_cType;
+  if (binder.tensorResultTypeAtIndex(yTy, 0) ||
+      binder.tensorResultTypeAtIndex(Y_hType, 1) ||
+      binder.tensorResultTypeAtIndex(Y_cType, 2)) {
+    return rewriter.notifyMatchFailure(binder.op,
+                                       "At least one outputs must be present");
+  }
+  Value X;
+  if (binder.tensorOperandAtIndex(X, 0))
+    return rewriter.notifyMatchFailure(binder.op,
+                                       "Missing required input tensor X");
+  Value W;
+  if (binder.tensorOperandAtIndex(W, 1))
+    return rewriter.notifyMatchFailure(binder.op,
+                                       "Missing required input tensor W");
+  Value R;
+  if (binder.tensorOperandAtIndex(R, 2))
+    return rewriter.notifyMatchFailure(binder.op,
+                                       "Missing required input tensor R");
+  int64_t hidden_size;
+  if (binder.s64IntegerAttr(hidden_size, "hidden_size"))
+    return rewriter.notifyMatchFailure(
+        binder.op, "Missing required attribute hidden_size");
+
+  auto xTy = cast<ValueTensorType>(X.getType());
+  auto wTy = cast<ValueTensorType>(W.getType());
+  Value B;
+  if (binder.tensorOperandAtIndex(B, 3)) {
+    B = b.create<AtenZerosOp>(W.getType(), W);
+  }
+
+  llvm::SmallVector<std::string> activationsList;
+  if (binder.stringArrayAttr(activationsList, "activations"))
+    return rewriter.notifyMatchFailure(
+        binder.op, "Missing required attribute; activations");
+
+  LstmActivations activations;
+  if (activationsList.size() != 3 && activationsList.size() != 0)
+    return rewriter.notifyMatchFailure(
+        binder.op, "Either empty or 3 activations are supported but " +
+                       std::to_string(activationsList.size()) +
+                       " are provided.");
+
+  if (activationsList.size() == 3) {
+    activations.f = activationsList[0];
+    activations.g = activationsList[1];
+    activations.h = activationsList[2];
+  } else {
+    activations.f = "Sigmoid";
```

Comment:
Rather than `if() { ... } else { ... }` I would just set to the defaults then change if norm. This interprets having a default and setting an explicit version.